src/Build/Construction/Solution/SolutionFile.cs

```diff
@@ -601,15 +605,47 @@ internal void ParseSolution()
                     }
                 }
 
-                // Update the hash table with this unique name
-                bool didntAlreadyExist = projectsByUniqueName.Add(uniqueName);
+                // Detect collision caused by unique name's normalization
+                if (projectsByUniqueName.ContainsKey(uniqueName))
+                {
+                    // Is normalization occured in the current project?
+                    if (uniqueName != proj.ProjectName)
+                    {
+                        // Generates a new unique name
+                        string tempUniqueName = $"{uniqueName}_{proj.GetProjectGuidWithoutCurlyBrackets()}";
+                        proj.UpdateUniqueProjectName(tempUniqueName);
+                        uniqueName = tempUniqueName;
+                    }
+                    // Is normalization occured in a previous project?
+                    else if (uniqueName != projectsByUniqueName[uniqueName].ProjectName)
+                    {
+                        var projTemp = projectsByUniqueName[uniqueName];
+
+                        // Generates a new unique name
+                        string tempUniqueName = $"{uniqueName}_{projTemp.GetProjectGuidWithoutCurlyBrackets()}";
+                        projTemp.UpdateUniqueProjectName(tempUniqueName);
+
+                        projectsByUniqueName.Remove(uniqueName);
```

Comment:
Is this scenario possible?
1) You have two projects A and C that are the same and a third, B, that happens to have the same unique name but is different.
2) A is checked first, then B. B has a different ProjectName than A, so A's uniqueName is updated here, and projectsByUniqueName now has references to both A and B.
3) C comes through third, and it is compared against _B_ which it's different from, so it generates a unique name that happens to match A's. C overwrites A, and the error isn't caught.

Reply:
Did you mean a scenario like the one below?

```
Microsoft Visual Studio Solution File, Format Version 12.00
# Visual Studio 15
VisualStudioVersion = 15.0.27130.2010
MinimumVisualStudioVersion = 10.0.40219.1
Project('{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}') = 'Project.Named.With.Dots', 'Project.Named.With.Dots.csproj', '{FC2889D9-6050-4D2E-B022-979CCFEEAAAC}'
EndProject
Project('{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}') = 'Project_Named_With_Dots', 'Project_Named_With_Dots.csproj', '{ED30D4A3-1214-410B-82BB-B61E5A9D05CA}'
EndProject
Project('{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}') = 'Project.Named.With.Dots', 'Project.Named.With.Dots.csproj', '{6185CC21-BE89-448A-B3C0-D1C27112E595}'
EndProject
Global
	GlobalSection(SolutionConfigurationPlatforms) = preSolution
		Release|Any CPU = Release|Any CPU
	EndGlobalSection
	GlobalSection(ProjectConfigurationPlatforms) = postSolution
		{FC2889D9-6050-4D2E-B022-979CCFEEAAAC}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{FC2889D9-6050-4D2E-B022-979CCFEEAAAC}.Release|Any CPU.Build.0 = Release|Any CPU
		{ED30D4A3-1214-410B-82BB-B61E5A9D05CA}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{ED30D4A3-1214-410B-82BB-B61E5A9D05CA}.Release|Any CPU.Build.0 = Release|Any CPU
		{6185CC21-BE89-448A-B3C0-D1C27112E595}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{6185CC21-BE89-448A-B3C0-D1C27112E595}.Release|Any CPU.Build.0 = Release|Any CPU
	EndGlobalSection
	GlobalSection(SolutionProperties) = preSolution
		HideSolutionNode = FALSE
	EndGlobalSection
	GlobalSection(ExtensibilityGlobals) = postSolution
		SolutionGuid = {C038ED6B-BFC1-4E50-AE2E-7993F6883D7F}
	EndGlobalSection
EndGlobal
```

If so, it will return the error: `The solution file has two projects named "Project.Named.With.Dots".`

Reply:
Ah, perfect! Missed that.